marimbabot_audio/scripts/note2midi.py

```diff
@@ -0,0 +1,177 @@
+            NoteOnset,
+            self.onset_event,
+            queue_size=500,
+            tcp_nodelay=True,
+        )
+
+        # self.pub_midi = rospy.Publisher(
+        #     "midi", Image, queue_size=1, tcp_nodelay=True
+        # )
+        self._init_plot()
+
+    def run(self):
+        rospy.spin()
+
+    def reset(self):
+        if os.path.exists(self.file_path):
+            os.remove(self.file_path)
+
+    def onset_event(self, msg: NoteOnset):
+
+        # parse the msg
+        start_time = msg.header.stamp.secs + msg.header.stamp.nsecs / 1000000000  # the system second, not start from zero.
+        note = msg.note
+        duration = msg.confidence
+
+        # load midi file
+        self.midi = MyPrettyMiDi(file_path=self.file_path)
+
+        # make sure the time counter start from zero
+        if self.time_first == 0:
+            self.time_first = start_time
+        note_start_time = start_time - self.time_first
+        note_end_time = note_start_time + duration
+
+        # append the note to the midi file under default instrument marimba
+
+        self.midi.append_note(
+            instrument_name="Marimba",
+            onset_note=note,
+            start_time=note_start_time,
+            end_time=note_end_time
+        )
+        # save to file
+        self.midi.save()
+        print(f"onset added :  onset_note:{note}  start_time:{start_time} end_time:{start_time + duration}")
+
+    ## For visualization
+    # This function is called periodically from FuncAnimation
+    def animate(self, i):
+        limit = 20
+        if os.path.exists(self.file_path):
+            notes = MyPrettyMiDi(file_path=self.file_path).pm.instruments[0].notes[-limit:]
+            pitchs = [note.pitch for note in notes]
+            start_time = [note.start for note in notes]
+            durations = [note.end-note.start for note in notes]
+            note_names = [pretty_midi.note_number_to_name(note.pitch) for note in notes]
+
+            # Limit x and y lists to 20 items
+            xs = start_time
+            ys = pitchs
+
+            # C4 = 60  C7 = 96
+            # Draw x and y lists
+            self.ax.clear()
+            for idx in range(len(xs)):
+                self.ax.bar(x=xs[idx], height=1, width=durations[idx], bottom=ys[idx] - 0.5, align="edge", color='b')
+
+            # Format plot
+            plt.xticks(rotation=45, ha='right')
+            plt.yticks(pitchs, note_names)
+            plt.subplots_adjust(bottom=0.30)
+            plt.title('marimba note over Time')
+            plt.ylabel('Notes')
+            plt.xlabel('Time')
+        else:
+            return
+
+    def _init_plot(self):
+        # Create figure for plotting
+        self.fig = plt.figure()
+        self.ax = self.fig.add_subplot(1, 1, 1)
+
+        # Set up plot to call animate() function periodically
+        ani = animation.FuncAnimation(self.fig, self.animate, fargs=(), interval=1000)
+        plt.show()
```

Comment:
Some feature idea: You can publish the buffer of the plot as an image ros topic instead of showing a matplotlib window.

Reply:
how about to do this part in next version. I put it in the code TODO

Reply:
@yunlong-wang-cn  could you create a new issue for that instead?

Reply:
done @Juphex